### DISCUSSION TITLE
Considerations regarding the WS update exchange

### DISCUSSION THREAD

**Discussion by pierrecdn** [zauberzeug/nicegui#3745] [Ideas / Feature Requests]
I started having a performance issue when a user is opening a dialog with many objects.
The browser rendering in itself is what causes most troubles, but actually I also identified something regarding the `update` payload exchanged on the websocket.

### `ui.markdown`

*Each* text which is contained in a `ui.markdown` block is wrapped into a dense and repeated `codehilite_css` key in the JSON, which is quite verbose:

<details><summary>Details</summary>
<pre>
pre { line-height: 125%; }
td.linenos .normal { color: inherit; background-color: transparent; padding-left: 5px; padding-right: 5px; }
span.linenos { color: inherit; background-color: transparent; padding-left: 5px; padding-right: 5px; }
td.linenos .special { color: #000000; background-color: #ffffc0; padding-left: 5px; padding-right: 5px; }
span.linenos.special { color: #000000; background-color: #ffffc0; padding-left: 5px; padding-right: 5px; }
.codehilite .hll { background-color: #ffffcc }
.codehilite .c { color: #3D7B7B; font-style: italic } /* Comment */
.codehilite .err { border: 1px solid #FF0000 } /* Error */
.codehilite .k { color: #008000; font-weight: bold } /* Keyword */
.codehilite .o { color: #666666 } /* Operator */
.codehilite .ch { color: #3D7B7B; font-style: italic } /* Comment.Hashbang */
.codehilite .cm { color: #3D7B7B; font-style: italic } /* Comment.Multiline */
.codehilite .cp { color: #9C6500 } /* Comment.Preproc */
.codehilite .cpf { color: #3D7B7B; font-style: italic } /* Comment.PreprocFile */
.codehilite .c1 { color: #3D7B7B; font-style: italic } /* Comment.Single */
.codehilite .cs { color: #3D7B7B; font-style: italic } /* Comment.Special */
.codehilite .gd { color: #A00000 } /* Generic.Deleted */
.codehilite .ge { font-style: italic } /* Generic.Emph */
.codehilite .ges { font-weight: bold; font-style: italic } /* Generic.EmphStrong */
.codehilite .gr { color: #E40000 } /* Generic.Error */
.codehilite .gh { color: #000080; font-weight: bold } /* Generic.Heading */
.codehilite .gi { color: #008400 } /* Generic.Inserted */
.codehilite .go { color: #717171 } /* Generic.Output */
.codehilite .gp { color: #000080; font-weight: bold } /* Generic.Prompt */
.codehilite .gs { font-weight: bold } /* Generic.Strong */
.codehilite .gu { color: #800080; font-weight: bold } /* Generic.Subheading */
.codehilite .gt { color: #0044DD } /* Generic.Traceback */
.codehilite .kc { color: #008000; font-weight: bold } /* Keyword.Constant */
.codehilite .kd { color: #008000; font-weight: bold } /* Keyword.Declaration */
.codehilite .kn { color: #008000; font-weight: bold } /* Keyword.Namespace */
.codehilite .kp { color: #008000 } /* Keyword.Pseudo */
.codehilite .kr { color: #008000; font-weight: bold } /* Keyword.Reserved */
.codehilite .kt { color: #B00040 } /* Keyword.Type */
.codehilite .m { color: #666666 } /* Literal.Number */
.codehilite .s { color: #BA2121 } /* Literal.String */
.codehilite .na { color: #687822 } /* Name.Attribute */
.codehilite .nb { color: #008000 } /* Name.Builtin */
.codehilite .nc { color: #0000FF; font-weight: bold } /* Name.Class */
.codehilite .no { color: #880000 } /* Name.Constant */
.codehilite .nd { color: #AA22FF } /* Name.Decorator */
.codehilite .ni { color: #717171; font-weight: bold } /* Name.Entity */
.codehilite .ne { color: #CB3F38; font-weight: bold } /* Name.Exception */
.codehilite .nf { color: #0000FF } /* Name.Function */
.codehilite .nl { color: #767600 } /* Name.Label */
.codehilite .nn { color: #0000FF; font-weight: bold } /* Name.Namespace */
.codehilite .nt { color: #008000; font-weight: bold } /* Name.Tag */
.codehilite .nv { color: #19177C } /* Name.Variable */
.codehilite .ow { color: #AA22FF; font-weight: bold } /* Operator.Word */
.codehilite .w { color: #bbbbbb } /* Text.Whitespace */
.codehilite .mb { color: #666666 } /* Literal.Number.Bin */
.codehilite .mf { color: #666666 } /* Literal.Number.Float */
.codehilite .mh { color: #666666 } /* Literal.Number.Hex */
.codehilite .mi { color: #666666 } /* Literal.Number.Integer */
.codehilite .mo { color: #666666 } /* Literal.Number.Oct */
.codehilite .sa { color: #BA2121 } /* Literal.String.Affix */
.codehilite .sb { color: #BA2121 } /* Literal.String.Backtick */
.codehilite .sc { color: #BA2121 } /* Literal.String.Char */
.codehilite .dl { color: #BA2121 } /* Literal.String.Delimiter */
.codehilite .sd { color: #BA2121; font-style: italic } /* Literal.String.Doc */
.codehilite .s2 { color: #BA2121 } /* Literal.String.Double */
.codehilite .se { color: #AA5D1F; font-weight: bold } /* Literal.String.Escape */
.codehilite .sh { color: #BA2121 } /* Literal.String.Heredoc */
.codehilite .si { color: #A45A77; font-weight: bold } /* Literal.String.Interpol */
.codehilite .sx { color: #008000 } /* Literal.String.Other */
.codehilite .sr { color: #A45A77 } /* Literal.String.Regex */
.codehilite .s1 { color: #BA2121 } /* Literal.String.Single */
.codehilite .ss { color: #19177C } /* Literal.String.Symbol */
.codehilite .bp { color: #008000 } /* Name.Builtin.Pseudo */
.codehilite .fm { color: #0000FF } /* Name.Function.Magic */
.codehilite .vc { color: #19177C } /* Name.Variable.Class */
.codehilite .vg { color: #19177C } /* Name.Variable.Global */
.codehilite .vi { color: #19177C } /* Name.Variable.Instance */
.codehilite .vm { color: #19177C } /* Name.Variable.Magic */
.codehilite .il { color: #666666 } /* Literal.Number.Integer.Long */pre { line-height: 125%; }
td.linenos .normal { color: #6e7681; background-color: #0d1117; padding-left: 5px; padding-right: 5px; }
span.linenos { color: #6e7681; background-color: #0d1117; padding-left: 5px; padding-right: 5px; }
td.linenos .special { color: #e6edf3; background-color: #6e7681; padding-left: 5px; padding-right: 5px; }
span.linenos.special { color: #e6edf3; background-color: #6e7681; padding-left: 5px; padding-right: 5px; }
.body--dark .codehilite .hll { background-color: #6e7681 }
.body--dark .codehilite .c { color: #8b949e; font-style: italic } /* Comment */
.body--dark .codehilite .err { color: #f85149 } /* Error */
.body--dark .codehilite .esc { color: #e6edf3 } /* Escape */
.body--dark .codehilite .g { color: #e6edf3 } /* Generic */
.body--dark .codehilite .k { color: #ff7b72 } /* Keyword */
.body--dark .codehilite .l { color: #a5d6ff } /* Literal */
.body--dark .codehilite .n { color: #e6edf3 } /* Name */
.body--dark .codehilite .o { color: #ff7b72; font-weight: bold } /* Operator */
.body--dark .codehilite .x { color: #e6edf3 } /* Other */
.body--dark .codehilite .p { color: #e6edf3 } /* Punctuation */
.body--dark .codehilite .ch { color: #8b949e; font-style: italic } /* Comment.Hashbang */
.body--dark .codehilite .cm { color: #8b949e; font-style: italic } /* Comment.Multiline */
.body--dark .codehilite .cp { color: #8b949e; font-weight: bold; font-style: italic } /* Comment.Preproc */
.body--dark .codehilite .cpf { color: #8b949e; font-style: italic } /* Comment.PreprocFile */
.body--dark .codehilite .c1 { color: #8b949e; font-style: italic } /* Comment.Single */
.body--dark .codehilite .cs { color: #8b949e; font-weight: bold; font-style: italic } /* Comment.Special */
.body--dark .codehilite .gd { color: #ffa198; background-color: #490202 } /* Generic.Deleted */
.body--dark .codehilite .ge { color: #e6edf3; font-style: italic } /* Generic.Emph */
.body--dark .codehilite .ges { color: #e6edf3; font-weight: bold; font-style: italic } /* Generic.EmphStrong */
.body--dark .codehilite .gr { color: #ffa198 } /* Generic.Error */
.body--dark .codehilite .gh { color: #79c0ff; font-weight: bold } /* Generic.Heading */
.body--dark .codehilite .gi { color: #56d364; background-color: #0f5323 } /* Generic.Inserted */
.body--dark .codehilite .go { color: #8b949e } /* Generic.Output */
.body--dark .codehilite .gp { color: #8b949e } /* Generic.Prompt */
.body--dark .codehilite .gs { color: #e6edf3; font-weight: bold } /* Generic.Strong */
.body--dark .codehilite .gu { color: #79c0ff } /* Generic.Subheading */
.body--dark .codehilite .gt { color: #ff7b72 } /* Generic.Traceback */
.body--dark .codehilite .g-Underline { color: #e6edf3; text-decoration: underline } /* Generic.Underline */
.body--dark .codehilite .kc { color: #79c0ff } /* Keyword.Constant */
.body--dark .codehilite .kd { color: #ff7b72 } /* Keyword.Declaration */
.body--dark .codehilite .kn { color: #ff7b72 } /* Keyword.Namespace */
.body--dark .codehilite .kp { color: #79c0ff } /* Keyword.Pseudo */
.body--dark .codehilite .kr { color: #ff7b72 } /* Keyword.Reserved */
.body--dark .codehilite .kt { color: #ff7b72 } /* Keyword.Type */
.body--dark .codehilite .ld { color: #79c0ff } /* Literal.Date */
.body--dark .codehilite .m { color: #a5d6ff } /* Literal.Number */
.body--dark .codehilite .s { color: #a5d6ff } /* Literal.String */
.body--dark .codehilite .na { color: #e6edf3 } /* Name.Attribute */
.body--dark .codehilite .nb { color: #e6edf3 } /* Name.Builtin */
.body--dark .codehilite .nc { color: #f0883e; font-weight: bold } /* Name.Class */
.body--dark .codehilite .no { color: #79c0ff; font-weight: bold } /* Name.Constant */
.body--dark .codehilite .nd { color: #d2a8ff; font-weight: bold } /* Name.Decorator */
.body--dark .codehilite .ni { color: #ffa657 } /* Name.Entity */
.body--dark .codehilite .ne { color: #f0883e; font-weight: bold } /* Name.Exception */
.body--dark .codehilite .nf { color: #d2a8ff; font-weight: bold } /* Name.Function */
.body--dark .codehilite .nl { color: #79c0ff; font-weight: bold } /* Name.Label */
.body--dark .codehilite .nn { color: #ff7b72 } /* Name.Namespace */
.body--dark .codehilite .nx { color: #e6edf3 } /* Name.Other */
.body--dark .codehilite .py { color: #79c0ff } /* Name.Property */
.body--dark .codehilite .nt { color: #7ee787 } /* Name.Tag */
.body--dark .codehilite .nv { color: #79c0ff } /* Name.Variable */
.body--dark .codehilite .ow { color: #ff7b72; font-weight: bold } /* Operator.Word */
.body--dark .codehilite .pm { color: #e6edf3 } /* Punctuation.Marker */
.body--dark .codehilite .w { color: #6e7681 } /* Text.Whitespace */
.body--dark .codehilite .mb { color: #a5d6ff } /* Literal.Number.Bin */
.body--dark .codehilite .mf { color: #a5d6ff } /* Literal.Number.Float */
.body--dark .codehilite .mh { color: #a5d6ff } /* Literal.Number.Hex */
.body--dark .codehilite .mi { color: #a5d6ff } /* Literal.Number.Integer */
.body--dark .codehilite .mo { color: #a5d6ff } /* Literal.Number.Oct */
.body--dark .codehilite .sa { color: #79c0ff } /* Literal.String.Affix */
.body--dark .codehilite .sb { color: #a5d6ff } /* Literal.String.Backtick */
.body--dark .codehilite .sc { color: #a5d6ff } /* Literal.String.Char */
.body--dark .codehilite .dl { color: #79c0ff } /* Literal.String.Delimiter */
.body--dark .codehilite .sd { color: #a5d6ff } /* Literal.String.Doc */
.body--dark .codehilite .s2 { color: #a5d6ff } /* Literal.String.Double */
.body--dark .codehilite .se { color: #79c0ff } /* Literal.String.Escape */
.body--dark .codehilite .sh { color: #79c0ff } /* Literal.String.Heredoc */
.body--dark .codehilite .si { color: #a5d6ff } /* Literal.String.Interpol */
.body--dark .codehilite .sx { color: #a5d6ff } /* Literal.String.Other */
.body--dark .codehilite .sr { color: #79c0ff } /* Literal.String.Regex */
.body--dark .codehilite .s1 { color: #a5d6ff } /* Literal.String.Single */
.body--dark .codehilite .ss { color: #a5d6ff } /* Literal.String.Symbol */
.body--dark .codehilite .bp { color: #e6edf3 } /* Name.Builtin.Pseudo */
.body--dark .codehilite .fm { color: #d2a8ff; font-weight: bold } /* Name.Function.Magic */
.body--dark .codehilite .vc { color: #79c0ff } /* Name.Variable.Class */
.body--dark .codehilite .vg { color: #79c0ff } /* Name.Variable.Global */
.body--dark .codehilite .vi { color: #79c0ff } /* Name.Variable.Instance */
.body--dark .codehilite .vm { color: #79c0ff } /* Name.Variable.Magic */
.body--dark .codehilite .il { color: #a5d6ff } /* Literal.Number.Integer.Long */
</pre>
</details> 

Do you think it would be possible to factorize this? I divided the payload by 10 (from 2.5MB to 250KB) just by removing ui.markdown.

### Default values of objects

All objects (especially noticed that for `events`) are sent with their default values being set. 
Example:
```
      "events": [
        {
          "listener_id": "7ff6aa4d-5c36-46bc-b0ab-47922212e82e",
          "type": "update:modelValue",
          "specials": [],
          "modifiers": [],
          "keys": [],
          "args": [
            null
          ],
          "throttle": 0,
          "leading_events": true, 
          "trailing_events": true, 
          "js_handler": null
        }
      ]  
```

Wouldn't it be convenient to only provide what is actually overriden?

Thanks for your input on this, compression does a good job with this but in the end it still consume more on the network and load more objects in memory on client side.

**Comment by falkoschindler** [comment]
Thanks for sharing your thoughts on the websocket payload size, @pierrecdn!

**`ui.markdown`**

Oh wow! Actually I wasn't aware of this huge waste of traffic. Sure, since we use a Python module for generating CSS for syntax highlighting, we need to somehow send it to the client. But it is the same code for every element and takes up almost 12KB each time. We should definitely find a better way. Maybe we can move it into a separate JS or CSS file, which is transmitted only once and cached in the browser. I'll think about it.

**`events`**

You are right. We could shave off some bytes by omitting empty lists and possibly other default values. The code gets a bit more complex, because frontend _and_ backend need to know the defaults, and constructing and reconstructing the payload becomes more challenging. But it is definitely possible. We did something very similar in #2702 for the element dictionary, but left the `events` list unchanged because only a tiny fraction of elements do have events registered. Therefore optimizing this data structure might only marginally improve the overall payload size.

**Comment by pierrecdn** [comment]
Yes, I really feel the `ui.markdown` is fundamentally what's the best to fix (and probably is easier than addressing the other part?).

--

A bit off-topic, but regarding performances overall, what do you suggest to analyze bottlenecks?

Typically, I have my browser window which is frozen, then ends up closing the WS connection and all opened dialogs.
I increased `reconnect_timeout` and `binding_refresh_interval` to higher values, but it is still unable to render correctly. The number of objects loaded is very large.

Snapshots #3 displayed here is when dialog is opened, while #1 is when it is closed.
Would you have a hint to drill-down to something actionable?

![image](https://github.com/user-attachments/assets/d3c6bee2-1124-4175-b7e6-409558970ad2)

**Reply by pierrecdn** [reply]
About :arrow_up:, collected more info thanks to firefox devtools and `about:processes` that allows to profile threads of a given process/tab.
Ended up with a call stack, then had to set `prod_js=False` in nicegui to get non-minified Vue.js output.

![image](https://github.com/user-attachments/assets/2efc281c-c8ae-445a-bfed-22d6836f3155)

Now I'm a bit stuck with Quasar/Vue things, not sure where I'm going, I feel I could be in [such kind of issue](https://github.com/quasarframework/quasar/issues/16024).
